### PR TITLE
tests: use pytest fixtures, unittest helpers when possible

### PR DESCRIPTION
**Pull request by Kyle-Verhoog**
This patch unifies all span fetching in tests to go through pytest fixtures, unittest helpers or the `DummyTracer` instance directly. The goal is to minimize the dependence on the implementation detail of where/how spans are collected.

Essentially:

```python
class TestCase(TracerTestCase):
    def test(self):
        ...
        spans = self.tracer.writer.pop()

# becomes

class TestCase(TracerTestCase):
    def test(self):
        ...
        spans = self.pop_spans()
```

```python
def test(self, tracer):
        ...
        spans = tracer.writer.pop()

# becomes

def test(self, tracer, test_spans):
        ...
        spans = test_spans.pop()
```

```python
tracer = DummyTracer()
tracer.writer = DummyWriter()
spans = tracer.writer.pop() 

# becomes

tracer = DummyTracer()
spans = tracer.pop()
```

### CODE REVIEW

File: tests/contrib/aiopg/py35/test.py [hunk 1]
```diff
@@ -52,7 +52,7 @@ async def _test_cursor_ctx_manager(self):
             assert len(rows) == 1
             assert rows[0][0] == 'blah'
 
-        spans = tracer.writer.pop()
+        spans = tracer.pop()
```
Comment (Yun-Kim) [inline tests/contrib/aiopg/py35/test.py:55]: Question: why do we use `tracer.pop()` here compared to `self.pop_spans()`? Is there a difference between using the tracer's method to pop or the TestCase method to pop?

Comment (Kyle-Verhoog) [reply tests/contrib/aiopg/py35/test.py:55]: Good catch! This one can be `self.pop_spans()`. I had missed converting this one over. There really isn't a difference but by using the unittest style one we get consistency and avoid using the not obvious `tracer.pop()` proxy.

**Review by Yun-Kim — APPROVED** [review]
Wow... that was a long one 😭 I added a couple suggestions to use `self.pop_spans()` when applicable, but these are trivial changes (no change to result). I had 2 points:
1. It seems like the pytest fixtures (specifically `test_spans`) are used quite frequently across all integration test modules, which results in duplicate code blocks everywhere for these fixtures. I wonder if we should combine and refactor these common pytest fixtures into a single utilities file for our test directory?
2. I feel like we should have some guidelines when to use `self.pop_spans()` vs. `self.tracer.pop()` vs `tracer.pop()`. I understand some testcase classes don't inherit off of our TracerTestCase class which necessitates using the tracer object directly, but I'm not too sure what the different use cases are for the former two. For the sake of consistency should we use one over the other?

**Comment by Kyle-Verhoog** [comment]
> It seems like the pytest fixtures (specifically test_spans) are used quite frequently across all integration test modules, which results in duplicate code blocks everywhere for these fixtures. I wonder if we should combine and refactor these common pytest fixtures into a single utilities file for our test directory?

Yep I noticed this too. There are global/top level `tracer` and `test_spans` fixtures in `tests/conftest.py`. The issue I think is that some integrations override the `tracer` fixture to work in the patching logic. It could be that pytest just magically picks out the root level `test_spans` fixture and uses the overriding `tracer` fixture with it. I'll give it a shot to see how it works.

> I feel like we should have some guidelines when to use self.pop_spans() vs. self.tracer.pop() vs tracer.pop(). I understand some testcase classes don't inherit off of our TracerTestCase class which necessitates using the tracer object directly, but I'm not too sure what the different use cases are for the former two. For the sake of consistency should we use one over the other?

Yeah I think `self.pop_spans()` should be used over `self.tracer.pop()` since the latter, as @P403n1x87 pointed out, strays from the actual tracer API which is not intuitive and can be confusing for someone new to the codebase.

That said, having some testing guidelines would be great! We could add it to the integration PR checklist, WDYT?

**Comment by Yun-Kim** [comment]
> Yep I noticed this too. There are global/top level `tracer` and `test_spans` fixtures in `tests/conftest.py`. The issue I think is that some integrations override the `tracer` fixture to work in the patching logic. It could be that pytest just magically picks out the root level `test_spans` fixture and uses the overriding `tracer` fixture with it. I'll give it a shot to see how it works.

Sounds good! Please let me know if it works out :) 

> Yeah I think `self.pop_spans()` should be used over `self.tracer.pop()` since the latter, as @P403n1x87 pointed out, strays from the actual tracer API which is not intuitive and can be confusing for someone new to the codebase.

> That said, having some testing guidelines would be great! We could add it to the integration PR checklist, WDYT?

Agreed! Definitely support adding guidelines to encourage using `self.pop_spans()` and using the `TracerTestCase` in testing.

**Review by jd — APPROVED** [review]

**Review by Yun-Kim — APPROVED** [review]

**Comment by Kyle-Verhoog** [comment]
Added an entry to the checklist in https://github.com/DataDog/dd-trace-py/pull/2127

**Pull request merged**